File: stack-vs-queue/Queue.py
class Queue:

    def __init__(self, capacity):
        self.front = self.size = 0
        self.rear = capacity - 1
        self.queue = [None] * capacity
        self.capacity = capacity

    def is_full(self):
        return self.size == self.capacity

    def is_empty(self):
        return self.size == 0

    def enqueue(self, element):
        if self.is_full():
            return False

        self.rear = (self.rear + 1) % (self.capacity)
        self.queue[self.rear] = element
        self.size = self.size + 1

    def dequeue(self):
        if self.is_empty():
            return False

        element = self.queue[self.front]
        self.front = (self.front + 1) % (self.capacity)
        self.size = self.size - 1
        return element

    def queue_front(self):
        if self.is_empty():
            return False

        return self.queue[self.front]

    def queue_rear(self):
        if self.is_empty():
            return False

        return self.queue[self.rear]

File: stack-vs-queue/java/Queue.java
public class Queue
{
    private int front, rear, size;
    int capacity;
    int[] queue;

    public Queue()
    {
        this(1000);
    }

    public Queue(int capacity)
    {
        this.capacity = capacity;
        this.front = this.size = 0;
        this.rear = capacity -1;
        this.queue = new int[this.capacity];
    }

    public boolean isFull()
    {
        return this.size == this.capacity;
    }

    public boolean isEmpty()
    {
        return this.size == 0;
    }

    public void enqueue(int element)
    {
        if (isFull())
            return;

        this.rear = (this.rear++) % this.capacity;
        this.queue[this.rear] = element;
        this.size++;
    }

    public int dequeue()
    {
        if (isEmpty())
            return Integer.MIN_VALUE;

        int element = this.queue[this.front];
        this.front = (this.front++) % this.capacity;
        this.size--;
        return element;
    }

    public int front()
    {
        if (isEmpty())
            return Integer.MIN_VALUE;

        return this.queue[this.front];
    }

    public int rear()
    {
        if (isEmpty())
            return Integer.MIN_VALUE;

        return this.queue[this.rear];
    }
}
